### ISSUE TITLE
Proxy issue for Flyingpigeon component on Ouranos server

### ISSUE THREAD

**Issue by tomLandry** [CanadianClimateDataPortal/Canadian-Climate-Data-Portal#41]
Here, we recapitulate critical issues from PAVICS backlog that are in scope for CCDP.

- [ ]  Gateway Error https://github.com/Ouranosinc/pavics-sdi/issues/93
- [ ]  Access to THREDDS from console https://github.com/Ouranosinc/pavics-sdi/issues/90
- [x] CANARIE compliance https://github.com/Ouranosinc/pavics-sdi/issues/84

**Comment by tomLandry** [comment]
@huard I had this issue to track critical from PAVICS, pertaining to CCDP. From this end, CANARIE compliance and deployment of latest version looks ok. I had Gateway error and THREDDS from console too, unsure if there is a side effect to the Portal project. Another way to see it if that helps us down the way with a better infrastructure or capabilities. 

Comments?

**Comment by huard** [comment]
The THREDDS issue would need to be tested against a server requiring identification. The PAVICS thredds is open. 
The gateway error is I think the most critical.

**Comment by dbyrns** [comment]
PAVICS thredds is open by choice using magpie settings. For testing purpose permissions can be set to keep everything open but a single test folder.

**Comment by huard** [comment]
Good idea. 
@tlvu @mourad Could you create a "/testdata/secure/" folder on the thredds server with a test json and netcdf file ? This folder would be closed to anonymous users, but open to the "authtest" user with password 12345. We would use this for testing.

**Comment by tlvu** [comment]
Tested locking down testdata/secure for anonymous and allowed access to authuser worked.  It's the PAVICS on my local machine.  I'll let @mourad reproduce the config to the prod server (It's not clear to me which prod server we want this on).

Mapgie anonymous group settings:

![magpie-group-anonymous-settings](https://user-images.githubusercontent.com/11966697/52968105-3e8d8700-337a-11e9-9af8-d79f981daffe.png)

Magpie authtest user settings:

![magpie-user-authtest-settings](https://user-images.githubusercontent.com/11966697/52968147-56650b00-337a-11e9-80c9-325494532fae.png)

anonymous works for testdata/flyingpigeon
```
$ curl https://lvu.ouranos.ca/thredds/catalog/birdhouse/testdata/flyingpigeon/catalog.html --insecure --silent | head
<!DOCTYPE html PUBLIC '-//W3C//DTD HTML 4.01 Transitional//EN'
'http://www.w3.org/TR/html4/loose.dtd'>
<html>
<head>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><title>Catalog http://lvu.ouranos.ca:8083/thredds/catalog/birdhouse/testdata/flyingpigeon/catalog.html</title>
<link rel='stylesheet' href='/thredds/' type='text/css' >
</head>
<body><h1><img src='/thredds/threddsIcon.gif' alt='TDS' align='left' valign='top'>
Catalog http://lvu.ouranos.ca:8083/thredds/catalog/birdhouse/testdata/flyingpigeon/catalog.html</h1><HR size='1' noshade='noshade'><table width='100%' cellspacing='0' cellpadding='5' align='center'>
<tr>
```
anonymous not allowed to access testdata/secure
```
$ curl https://lvu.ouranos.ca/thredds/catalog/birdhouse/testdata/secure/catalog.html --insecure
<?xml version="1.0" encoding="utf-8"?>
<ExceptionReport version="1.0.0"
xmlns="http://www.opengis.net/ows/1.1"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd">
<Exception exceptionCode="NoApplicableCode" locator="AccessForbidden">
<ExceptionText>Not authorized to access this resource.</ExceptionText>
</Exception>
</ExceptionReport>
```

login with user authtest from magpie, extract the cookie from the browser
user authtest is allowed to access testdata/secure
```
$ curl https://lvu.ouranos.ca/thredds/catalog/birdhouse/testdata/secure/catalog.html --insecure --silent --cookie 'auth_tkt=c6ac5fcea64af606786b026ce067ee18c725f771bace5f3cff8ec724f8340e1f65f58f3148e8b3f1f49438f2560aa3111077109a9c4bccb58697b60f4739b6745c6ae0af5!userid_type:int' | head
<!DOCTYPE html PUBLIC '-//W3C//DTD HTML 4.01 Transitional//EN'
'http://www.w3.org/TR/html4/loose.dtd'>
<html>
<head>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><title>Catalog http://lvu.ouranos.ca:8083/thredds/catalog/birdhouse/testdata/secure/catalog.html</title>
<link rel='stylesheet' href='/thredds/' type='text/css' >
</head>
<body><h1><img src='/thredds/threddsIcon.gif' alt='TDS' align='left' valign='top'>
Catalog http://lvu.ouranos.ca:8083/thredds/catalog/birdhouse/testdata/secure/catalog.html</h1><HR size='1' noshade='noshade'><table width='100%' cellspacing='0' cellpadding='5' align='center'>
<tr>
```

**Comment by huard** [comment]
Good, the next step is to make sure this access works from a python console. To be clear, something this needs to work (see http://xarray.pydata.org/en/stable/io.html?highlight=password#opendap): 
```
import xarray as xr
import requests

session = requests.Session()
session.auth = ('username', 'password')

store = xr.backends.PydapDataStore.open('http://example.com/data',
                                        session=session)
ds = xr.open_dataset(store)
```

The other thing that we must check is running a wps process from the console (birdy) that requires a secured file. I suggest we use bird-house/emu as the test server.

**Comment by moulab88** [comment]
The config is now applied on Medus for testing.

**Comment by moulab88** [comment]
emu on medus:
https://medus.ouranos.ca/twitcher/ows/proxy/emu/wps?service=WPS&version=1.0.0&request=GetCapabilities
http://medus.ouranos.ca:8094/wps?service=WPS&version=1.0.0&request=GetCapabilities

**Comment by tomLandry** [comment]
@huard can you rate the priority of this issue for March 31st? Is it blocking to your activities?

**Comment by huard** [comment]
@tomLandry I believe that the issue has so far mostly affected flyingpigeon. The priority would be to confirm whether or not the Finch deployment on PAVICS suffers from the same bad gateway issue. If not, then the issue would be low priority. If it does, it would be high priority.

**Comment by tomLandry** [comment]
Thanks. I then rename the issue as it is specific to your deployment. We report no proxy issue on https://finch.crim.ca/api#/ and we intend to use that endpoint for analytics. I move back the issue in the backlog and will look at it again April 1st. Just wave if priority changes on your side.

**Comment by rod-glover** [comment]
Just had a look at your very nice OpenAPI / Swagger API spec. Thought you might be interested to know about, if you do not already, the Python package [Connexion](https://github.com/zalando/connexion). I haven't exercised it very hard yet, but it looks useful and is so far holding up well.

**Comment by davidcaron** [comment]
@rod-glover I'm using connexion for another project and it's fantastic. It reduces boilerplate a lot.

For this project though, we use pyramid, cornice and colander.

**Comment by rod-glover** [comment]
@davidcaron , I'd be interested sometime to talk about your experience with pyramid et. al. But I know now is definitely not the time. One day...

**Comment by tomLandry** [comment]
@rod-glover Yeah... this week is pretty crazy indeed. We're looking at when that could happen during the summer. Also, I would get your attention on another of our components, https://github.com/crim-ca/weaver, that we actively develop inside OGC's innovation program. That could become pretty cornerstone for medium-term projects - including you guys at PCIC!

**Comment by rod-glover** [comment]
@tomLandry , yes, I had a note to look into weaver. Thanks for the link to it. It looks very ambitious, and very interesting. I'm going to set aside some time to look at it more deeply fairly soon. Are the [Execute](https://finch.crim.ca/api#/Execute) API endpoints an interface to whatever system uses Weaver?

**Comment by dbyrns** [comment]
@rod-glover. If I understand correctly your question, yes the API you refered to is also the API of Weaver and is the interface that should be used by clients. This API provide access to WPS 1.0 providers like Finch or Flyingpigeon, but also to some ESGF processes via CWT and the ADES service developped in OGC's testbed that let execute Docker applications. On top of that it can execute workflows based on any of these processes.

**Comment by tomLandry** [comment]
Nicely put David. And at some point, in "federated cyberinfratructures", we'd like to be able to use data and processes offered in nodes deployed at PCIC. The ADES part - Application Deployment and Execution Service - does allow dynamic deployment to endpoints of these "applications".
